### DISCUSSION TITLE
Store static token hashed

### DISCUSSION THREAD

**Discussion by wc-matteo** [directus/directus#5744] [Feature Requests]
Shouldn't the static token be stored as a hash since it's essentially a password?

**Comment by rijkvanzanten** [comment]
If we would store it hashed, there would be no way to know what user goes with the provided token, as the api would have to fetch every user in the platform, and compare the hash one by one until it finds one 🤔

**Reply by wc-matteo** [reply]
maybe I'm missing something but... if you store the token hash instead of the token in the user data, then when you receive an api request you hash the token you receive and you do a lookup in the db to find if/to whom it belongs to. Wouldn't it work? And wouldn't it be the same as using a plain-text token, only with the addition of the hash part for the token received?

**Reply by rijkvanzanten** [reply]
The output hash is different every time, so you can't re-hash it first, and then use the result to compare

**Reply by wc-matteo** [reply]
Then it shouldn't be different xD It should work like a password hash, since it's essentially a password

**Reply by rijkvanzanten** [reply]
> Then it shouldn't be different xD

I'm confused.. It is? Every time you hash the same string, the hash is different. Therefore, we can't say "Get user where token hash is xxx", as xxx is different every time.

Take the following example:

```js
const argon = require('argon2');

async function run() {
    for (let i = 0; i < 5; i++) {
        console.log(await argon.hash('example'));
    }
}

run();
```

The output will be something like

```
$argon2i$v=19$m=4096,t=3,p=1$2vqWDPn+3+Fg7Gmr7bz4qg$KF4yoZHjjKQ/i0m5bAZcRFjbdZeMCX7q6FRS6NKZKxg
$argon2i$v=19$m=4096,t=3,p=1$1ywFbT9AsbB8AGDj5/ZTmw$UsdwX0jGHzQfS/v3/nSHN6JRhyp0ocHF0cxxWVz0IWk
$argon2i$v=19$m=4096,t=3,p=1$Be2qjwLy4l5OFB4JFt8/+g$eq1s4gmuuZAwTajGgyPh8zj/hy9A8wX/ROFFd728PB8
$argon2i$v=19$m=4096,t=3,p=1$tirWPF6deIhsrnX+qJUMYw$h56VVX5rbBUf0V+SYMdde4Kp53a/sM1BD5BveqAEr3I
$argon2i$v=19$m=4096,t=3,p=1$6OMKBItHmMZ2jqWrKW/J9Q$Mp7V6njow8CB0yInLP17uoPiYGjXO2SkEjJh2ccfFp4
```

The hash is different every generation for the same input string. If we were to store item 0 in this list to the DB, we can't use item 1 to find the record that contains item 0.

> It should work like a password hash, since it's essentially a password

It's not the same flow though. When you login with email/password, you're explicitly providing the unique identifier of the account (email in this case) which is used to fetch the stored hash, which can then be verified.

**Reply by wc-matteo** [reply]
> which is used to fetch the stored hash, which can then be verified

after you fetch the stored hash, how do you verify it, tho?

**Reply by wc-matteo** [reply]
the issue is, since the token is not required to be unique among all the users, there could be multiple tokens with the same value

**Reply by rijkvanzanten** [reply]
> after you fetch the stored hash, how do you verify it, tho?

We rely on argon2's provided `verify` function, which extracts the salt and parameters from the stored hash, and then re-hashes the input value with the known parameters to compare the output value.

> the issue is, since the token is not required to be unique among all the users, there could be multiple tokens with the same value

That sounds like an easy bug to fix! Directus recently started supporting `unique` constraints, so that should be enabled on the `directus_user.token` column as well 👍🏻

**Reply by wc-matteo** [reply]
> We rely on argon2's provided `verify` function, which extracts the salt and parameters from the stored hash, and then re-hashes the input value with the known parameters to compare the output value.

Oh I didn't think of the salt. Right. That's why the hash returned is always different for the same input value. Well what about not using the salt for the static tokens?

**Reply by wc-matteo** [reply]
> In cryptography, a salt is random data that is used as an additional input to a one-way function that hashes data, a password or passphrase. Salts are used to safeguard passwords in storage. Historically, only a cryptographic hash function of the password was stored on a system, but over time, additional safeguards were developed to protect against *duplicate or common passwords* being identifiable (as their hashes are identical). Salting is one such protection.
https://en.wikipedia.org/wiki/Salt_(cryptography)

"duplicate or common passwords" wouldn't be an issue with _unique_ static tokens

**Reply by rijkvanzanten** [reply]
Good point 👍🏻 

We'll have to do a bit of research to make sure there aren't any other unknown side-effects of [disabling the salt](https://github.com/ranisalt/node-argon2/wiki/Options#salt), and if there's any other things we have to tweak in argon's settings to make it usable and secure, but I'm down to make that switch as it surely beats having plain text in the db 👍🏻

**Reply by wc-matteo** [reply]
nice!

The alternative would be to pass to the api a user id together with the token. Then it would effectively work like a user/password system

**Reply by rijkvanzanten** [reply]
I think we should rely on a regular SHA256 hash for these tokens, as (like you correctly said) we don't need a salt, and argon2 has all sorts of failsafes in place to prevent us from setting saltLength to 0, or use a salt that's < 16 chars long 👍🏻 

That way we can also easily calculate the hash on use, then lookup the user based on the hash. This shouldn't be a breaking change, as we can set those hashes in a migration (and make the column unique in the process).

**Reply by rijkvanzanten** [reply]
@wc-matteo Would you like to take a swing at a PR? That's the fastest way to get this landed 🙂

**Reply by wc-matteo** [reply]
I'll see what I can do :)

**Reply by udf2457** [reply]
@rijkvanzanten @wc-matteo Instead of disabling the salt, how about a static system-wide salt for tokens ?

**Reply by rijkvanzanten** [reply]
@udf2457 As long as we have a deterministic hash, it should be possible to rehash the value on use, and then lookup the user based on the same hash output. Not as secure, but certainly better than plain text.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
